### ISSUE TITLE
Support changing the internal server name/id via Caddyfile

### ISSUE THREAD

**Issue by danlsgiga** [caddyserver/caddy#5322]
It doesn't seem to be possible to define the `server` names via Caddyfile for better identification through logs or metrics.

When using Caddyfile format, servers are named following a `srvX` pattern and when multiple sites are defined it becomes hard to identify which one it is referring to in logs or metrics

This is currently possible via the JSON [servers](https://caddyserver.com/docs/json/apps/http/servers/) keys.

**Comment by mholt** [comment]
What's your use case for needing the JSON key names when using Caddyfile? They are considered an implementation detail, and irrelevant when using Caddyfile. They're really just for your convenience of reading JSON config.

**Comment by danlsgiga** [comment]
I'm aware of https://github.com/caddyserver/caddy/issues/3784 which is kinda related. My use case is more related to logs / metrics where these are used as identifiers.

#### My use case

We are building a Caddy lb policy as a module to implement shuffle sharding internally. The way we are doing it is by defining a `main` site that has a `reverse_proxy` with upstreams being the same caddy instance for other purposes.

```
:8080 {
...
}

:8081 {
...
}

:8082 {
...
}

main {
  reverse_proxy {
    to localhost:8080 localhost:8081 localhost:8082
    lb_policy shuffle_shard <custom_module_options>
  }
}

```

The thing is when we instrument our metrics or logs, I get `srv0`, `srv1`, `srv2` and `srv3` as identifiers and those are hard to map to the right servers.

If I could, via Caddyfile, set a custom name to the server, these would be mapped to logs and metrics and I could easily filter based on that.

If there was a directive I could set inside each of these sites to replace the internal server name, that would be the easiest path from my limited understanding.

**Comment by WeidiDeng** [comment]
You can use python scripts to automate label changing, just using python subprocess to call caddy adapt and parse json config from subprocess stdout and modify config and feed it to caddy or save it to a file and let caddy run from it.

Parsing Caddyfile is a lot more difficult than it seems.

**Comment by francislavoie** [comment]
The problem with this is that servers are not 1:1 with the Caddyfile. There's risk that attempting to change the name could apply to more than one server, which would cause an invalid config because of duplicate JSON keys.

**Comment by WeidiDeng** [comment]
> The problem with this is that servers are not 1:1 with the Caddyfile. There's risk that attempting to change the name could apply to more than one server, which would cause an invalid config because of duplicate JSON keys.

That's why testing is important. Messing with raw json config is always a risk.

To be fair, using python or any other program languages to write json with built-in json library, there can't be duplicated keys.

**Comment by danlsgiga** [comment]
Would you be able to pinpoint in the code where the `srvX` logic is applied when using Caddyfile instead of raw JSON as the config adapter? I'm far from understanding if this is even possible given the feedback from y'all but I'd love to take a look at it!

**Comment by francislavoie** [comment]
The server name gets set here:

https://github.com/caddyserver/caddy/blob/c38a040e85a7d277299acaab39bdad171bc45675/caddyconfig/httpcaddyfile/httptype.go#L704

But the code to get to that point is very complex. There's a lot of sorting, organizing, disambiguating of the site blocks to get to "server pairings" and then it gets assembled.

**Comment by francislavoie** [comment]
I decided to look if there was actually any edgecase I could think of that wouldn't work. The only one I could actually confirm is a `servers { }` global option with no listener address, which obviously can't work since it means setting a name for all servers.

So I implemented it in #5323, it seems to work just fine. See the test for an example of how it would look.

**Comment by francislavoie** [comment]
Actually, one case it won't cover (and can't cover, as-is) is naming the HTTP server created via Auto-HTTPS. That's cause the server name config doesn't persist past adapting the config, and Auto-HTTPS happens at runtime, from the JSON config. So if you really care to name the HTTP server, then you'd need to create it yourself ahead of time with an empty `:80` (or `http://`) site block and rename it with `servers :80 { name http }` for example.
